algo/uidlist.go

```diff
@@ -105,46 +106,50 @@ func IntersectCompressedWithBin(dec *codec.Decoder, q []uint64, o *[]uint64) {
 	}
 
 	// Pick the shorter list and do binary search
-	if ld < lq {
+	if ld <= lq {
 		for {
 			blockUids := dec.Uids()
 			if len(blockUids) == 0 {
 				break
 			}
-			IntersectWithBin(blockUids, q, o)
-			lastUid := blockUids[len(blockUids)-1]
-			qidx := sort.Search(len(q), func(idx int) bool {
-				return q[idx] >= lastUid
-			})
-			if qidx >= len(q) {
+			if ld*linVsBinRatio < len(q) {
+				q = q[IntersectWithBin(blockUids, q, o):]
+			} else {
+				// For small enough difference between two arrays, we should just
+				// do lin intersect
+				_, off := IntersectWithLin(blockUids, q, o)
+				q = q[off:]
+			}
+			if len(q) == 0 {
 				return
 			}
-			q = q[qidx:]
 			dec.Next()
 		}
 		return
 	}
 
-	var uids []uint64
-	for _, u := range q {
+	uids := dec.Uids()
+	qidx := 0
+	for {
+		if qidx >= len(q) {
+			return
+		}
+		u := q[qidx]
 		if len(uids) == 0 || u > uids[len(uids)-1] {
-			uids = dec.Seek(u, codec.SeekStart)
+			if lq*linVsBinRatio < ld {
```

Comment:
fix this condition

Reply:
I tried updating the condition, but it made performance much worse. I am guessing its because that the decision to do binary or linear really depends on the total size of the arrays. If the array ratio is too high, then the numbers would be far apart.